<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>PhosFake Parameters</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
</style>


<script src="Parameters_files/libs/clipboard/clipboard.min.js"></script>
<script src="Parameters_files/libs/quarto-html/quarto.js"></script>
<script src="Parameters_files/libs/quarto-html/popper.min.js"></script>
<script src="Parameters_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="Parameters_files/libs/quarto-html/anchor.min.js"></script>
<link href="Parameters_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="Parameters_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="Parameters_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="Parameters_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="Parameters_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">PhosFake Parameters</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="phosfake-pipeline-parameters" class="level1">
<h1>PhosFake Pipeline Parameters</h1>
<p>This document provides a detailed description of the parameters used in the PhosFake pipeline. PhosFake is a tool for generating an in-silico bottom-up LC-MS dataset from proteoforms. The pipeline simulates various stages of the experimental process, from the generation of ground truth data to mass spectrometry (MS) analysis and statistical testing. For more information about the project, visit <a href="https://github.com/veitveit/PhosFake">PhosFake on GitHub</a>.</p>
<section id="global-parameters" class="level2">
<h2 class="anchored" data-anchor-id="global-parameters">Global Parameters</h2>
<section id="parallel-computing" class="level3">
<h3 class="anchored" data-anchor-id="parallel-computing">Parallel Computing</h3>
<p><em>Param$Cores</em>: The number of CPU cores to use for parallel computing. Set to <code>1</code>. Using multiple cores can significantly speed up the computation.</p>
<p><em>Param$ClusterType</em>: The type of cluster to use for parallel computing. Currently set to <code>PSOCK</code>. The cluster type determines how parallel processing is managed.</p>
</section>
<section id="section" class="level3">
<h3 class="anchored" data-anchor-id="section"></h3>
</section>
</section>
<section id="experimental-design" class="level2">
<h2 class="anchored" data-anchor-id="experimental-design">Experimental Design</h2>
<section id="number-of-conditions" class="level3">
<h3 class="anchored" data-anchor-id="number-of-conditions">Number of Conditions</h3>
<p><em>Param$NumCond</em>: The number of different experimental conditions to be tested. For example, if you are comparing treated versus untreated samples, this value would be 2. This parameter is crucial for setting up the experimental design. Default set to <code>2</code>.</p>
</section>
<section id="number-of-replicates" class="level3">
<h3 class="anchored" data-anchor-id="number-of-replicates">Number of Replicates</h3>
<p><em>Param$NumReps</em>: The number of biological or technical replicates for each condition. Replicates are essential for statistical validity and reproducibility of the results. Ensuring adequate replication helps in reducing variability and increasing the reliability of the results. Default set to <code>5</code>.</p>
</section>
</section>
<section id="input-data-paths" class="level2">
<h2 class="anchored" data-anchor-id="input-data-paths">Input Data Paths</h2>
<section id="path-to-fasta-file" class="level3">
<h3 class="anchored" data-anchor-id="path-to-fasta-file">Path to FASTA File</h3>
<p><em>Param$PathToFasta</em>: The file path to the FASTA file containing the protein sequences. This file is essential for identifying and quantifying proteins in your samples. Accurate protein identification depends heavily on the quality and completeness of the FASTA file. Default set to <code>InputData/fasta_full_human.fasta</code>.</p>
</section>
<section id="path-to-protein-list" class="level3">
<h3 class="anchored" data-anchor-id="path-to-protein-list">Path to Protein List</h3>
<p><em>Param$PathToProteinList</em>: The file path to a list of proteins of interest, if any. This can be used to focus the analysis on a subset of proteins. Providing a specific list can streamline the analysis and enhance focus on a set of related proteoforms or biological pathways. Default set to <code>NULL</code>.</p>
</section>
</section>
<section id="ground-truth-proteoform-centric-data-generation" class="level2">
<h2 class="anchored" data-anchor-id="ground-truth-proteoform-centric-data-generation">Ground Truth Proteoform-Centric Data Generation</h2>
<section id="proteoform-ids-generation" class="level3">
<h3 class="anchored" data-anchor-id="proteoform-ids-generation">Proteoform IDs Generation</h3>
<p><em>Param$FracModProt</em>: The fraction of proteins selected to be modified. This parameter controls how many proteins will undergo post-translational modifications (PTMs). It helps in simulating realistic biological scenarios. Default set to <code>0.0</code>, indicating no proteins will be modified.</p>
<p><em>Param$PropModPerProt</em>: The proportion of modifiable proteins sampled for modifications. A value of <code>2</code> means twice as many proteoforms as the selected proteins will be generated. This ensures diversity in the generated proteoforms.</p>
<p><em>Param$PTMTypes</em>: The types of PTMs to be considered. Default set to <code>ph</code> for phosphorylation. This parameter allows for the inclusion of different types of PTMs in the simulation. For multiple parameter in the same simulation, use a list: <code>list(c("ph", "ac"))</code></p>
<p><em>Param$PTMTypesDist</em>: The distribution of PTM types. Default set to <code>1</code>, indicating 100% of the given (one) PTM type. This controls the relative frequency of each PTM type.</p>
<p><em>Param$PTMTypesMass</em>: The mass shifts associated with the PTM types. This can be <code>79.9663</code> for phosphorylation or multiple value in a list such as <code>list(c(79.9663, 42.0106))</code> . Accurate mass shifts determine the mz values of the peptides.</p>
<p><em>Param$PTMMultipleLambda</em>: The lambda parameter for the Poisson distribution of multiply modified proteins. Set to <code>0.5</code>, scaled to the number of possible PTM sites. This parameter models the frequency of multiple modifications on the same protein.</p>
<p><em>Param$ModifiableResidues</em>: The residues where each PTM can occur. For phosphorylation, these are usually serine (S), threonine (T), and tyrosine (Y). This parameter defines the possible modification sites on the proteins. Each PTM type requires a set of residues. For phosphorylation, usually use <code>list(ph = c("S", "T", "Y"))</code></p>
<p><em>Param$ModifiableResiduesDistr</em>: The relative distribution of modifiable residues. For phosphorylation, this can be set to <code>0.86</code> for serine, <code>0.13</code> for threonine, and <code>0.01</code> for tyrosine. This reflects the observed frequencies of modifications at these residues in real biological data.</p>
<p><em>Param$RemoveNonModFormFrac</em>: The fraction of modifiable proteins without non-modified forms. Set to <code>0</code>. This parameter controls the presence of non-modified forms of proteins in the simulation.</p>
</section>
<section id="proteoform-abudance-generation" class="level3">
<h3 class="anchored" data-anchor-id="proteoform-abudance-generation">Proteoform Abudance Generation</h3>
<p><em>Param$QuantColnames</em>: The column names for quantitative values, structured as <code>C_&lt;condition&gt;_R_&lt;replicate&gt;</code>for a clear and organized structure for the quantitative data.</p>
<p><em>Param$QuantNoise</em>: The general noise level for all quantitative values, given as the standard deviation of a normal distribution. Default set to <code>0.0</code>. This parameter simulates technical and biological variations in the quantitative data.</p>
<p><em>Param$DiffRegFrac</em>: The fraction of differentially regulated proteoforms. Default set to <code>0.0</code>. This parameter allows for the simulation of differential regulation, mimicking biological variability.</p>
<p><em>Param$DiffRegMax</em>: The maximum amplitude of difference for differentially regulated proteins on a log2 scale. Set to <code>3</code>. This parameter defines the range of fold changes for differentially regulated proteins.</p>
<p><em>Param$UserInputFoldChanges</em>: User-defined fold changes for regulated proteoforms. Set to <code>NULL</code>. This allows for custom input of fold changes if specific regulation patterns are desired.</p>
<p><em>Param$ThreshNAProteoform</em>: The threshold for removing quantitative values at the proteoform level. Default set to <code>-100</code>. This parameter filters out low-abundant proteoforms.</p>
<p><em>Param$AbsoluteQuanMean</em>: The mean value for absolute quantitation. Set to <code>30.5</code>. This parameter sets the baseline for absolute quantitation values.</p>
<p><em>Param$AbsoluteQuanSD</em>: The standard deviation for absolute quantitation. Set to <code>3.6</code>. This parameter models the variability in absolute quantitation.</p>
<p><em>Param$ThreshNAQuantileProt</em>: The quantile threshold for removing quantitative values. Set to <code>0.01</code>. This parameter ensures that a certain proportion of low-abundant proteoforms are cut off.</p>
</section>
</section>
<section id="enzymatic-digestion-parameters" class="level2">
<h2 class="anchored" data-anchor-id="enzymatic-digestion-parameters">Enzymatic Digestion Parameters</h2>
<section id="enzyme" class="level3">
<h3 class="anchored" data-anchor-id="enzyme">Enzyme</h3>
<p><em>Param$Enzyme</em>: The enzyme used for protein digestion. Currently set to <code>trypsin</code>. Trypsin is commonly used in proteomics for its specificity and efficiency in cleaving proteins. Other possible values are trypsin, trypsin.strict, chymotrypsin.h, chymotrypsin.l, pepsin.2, pepsin.1.3, lysC, argC.</p>
</section>
<section id="proportion-of-peptides-with-missed-cleavages" class="level3">
<h3 class="anchored" data-anchor-id="proportion-of-peptides-with-missed-cleavages">Proportion of Peptides with Missed Cleavages</h3>
<p><em>Param$PropMissedCleavages</em>: The additional proportion of peptides that have missed cleavages. Set to <code>1</code>. Missed cleavages occur when the enzyme does not cleave at all possible sites, which can affect peptide identification.</p>
</section>
<section id="maximum-number-of-missed-cleavages" class="level3">
<h3 class="anchored" data-anchor-id="maximum-number-of-missed-cleavages">Maximum Number of Missed Cleavages</h3>
<p><em>Param$MaxNumMissedCleavages</em>: The maximum number of missed cleavages per peptide. Set to <code>6</code>. This parameter defines the upper limit of missed cleavages allowed in the simulation.</p>
</section>
<section id="peptide-length-filter" class="level3">
<h3 class="anchored" data-anchor-id="peptide-length-filter">Peptide Length Filter</h3>
<p><em>Param$PepMinLength</em>: The minimum length of peptides to be considered. Set to <code>7</code>. This filter ensures that only peptides of a certain length are included, which is important for reliable identification.</p>
<p><em>Param$PepMaxLength</em>: The maximum length of peptides to be considered. Set to <code>30</code>. This parameter excludes excessively long peptides that might not be covered by the mass range.</p>
</section>
<section id="least-abundant-peptide-loss" class="level3">
<h3 class="anchored" data-anchor-id="least-abundant-peptide-loss">Least Abundant Peptide Loss</h3>
<p><em>Param$LeastAbundantLoss</em>: The percentage of least abundant peptides to be removed. Set to <code>0.0</code>. This parameter allows for the exclusion of peptides that are too low in abundance to be reliably quantified.</p>
</section>
</section>
<section id="sample-preparation-simulation-parameters" class="level2">
<h2 class="anchored" data-anchor-id="sample-preparation-simulation-parameters">Sample Preparation Simulation Parameters</h2>
<section id="enrichment-loss" class="level3">
<h3 class="anchored" data-anchor-id="enrichment-loss">Enrichment Loss</h3>
<p><em>Param$EnrichmentLoss</em>: The loss of phosphorylated peptides during enrichment. Set to <code>0.2</code>. This parameter models the inefficiencies in the enrichment process, where some peptides are lost.</p>
</section>
<section id="enrichment-efficiency" class="level3">
<h3 class="anchored" data-anchor-id="enrichment-efficiency">Enrichment Efficiency</h3>
<p><em>Param$EnrichmentEfficiency</em>: The efficiency of enrichment for phosphorylated peptides. Default set to <code>1</code>. This parameter reflects how effectively phosphorylated peptides are enriched relative to non-modified peptides, given by the adjusting the peptide intensities</p>
</section>
<section id="enrichment-protocol-noise" class="level3">
<h3 class="anchored" data-anchor-id="enrichment-protocol-noise">Enrichment Protocol Noise</h3>
<p><em>Param$EnrichmentNoise</em>: The noise introduced by the enrichment protocol. Set to <code>0.2</code>. This parameter adds variability to the enrichment process.</p>
</section>
</section>
<section id="mass-spectrometry-ms-run-parameters" class="level2">
<h2 class="anchored" data-anchor-id="mass-spectrometry-ms-run-parameters">Mass Spectrometry (MS) Run Parameters</h2>
<section id="detectability-filter" class="level3">
<h3 class="anchored" data-anchor-id="detectability-filter">Detectability filter</h3>
<p><em>Param$PercDetectability:</em> Removing the given percentage of peptides according to their predicted detectibility from PeptideRanger predictions (trained on ProteomicsDB). Default value is 0.25</p>
<p><em>Param$PercDetectedPep</em>: The percentage of peptides detected during the MS run. Set to <code>0.3</code>. This parameter models the efficiency of peptide detection by the MS instrument.</p>
</section>
<section id="detected-values" class="level3">
<h3 class="anchored" data-anchor-id="detected-values">Detected Values</h3>
<p><em>Param$PercDetectedVal</em>: The percentage of detected values in each replicate or condition. Set to <code>0.3</code>. This parameter affects the completeness of the quantitative data.</p>
</section>
<section id="intensity-dependence-weight" class="level3">
<h3 class="anchored" data-anchor-id="intensity-dependence-weight">Intensity-Dependence Weight</h3>
<p><em>Param$WeightDetectVal</em>: The weight for intensity dependence of non-detection, with a maximum of <code>40</code>. Currently set to <code>10</code>. This parameter models the likelihood of detection based on peptide intensity.</p>
</section>
<section id="ms-noise" class="level3">
<h3 class="anchored" data-anchor-id="ms-noise">MS Noise</h3>
<p><em>Param$MSNoise</em>: The noise level added by the MS instrument. Set to <code>0.5</code>. This parameter simulates the inherent noise in MS measurements.</p>
</section>
</section>
<section id="ms-search-parameters" class="level2">
<h2 class="anchored" data-anchor-id="ms-search-parameters">MS Search Parameters</h2>
<section id="wrong-identifications" class="level3">
<h3 class="anchored" data-anchor-id="wrong-identifications">Wrong Identifications</h3>
<p><em>Param$WrongIDs</em>: The percentage of wrong identifications in the MS search. Set to <code>0.01</code>. This parameter models the rate of incorrect peptide identifications due to errors in the database search process.</p>
</section>
<section id="wrong-localizations" class="level3">
<h3 class="anchored" data-anchor-id="wrong-localizations">Wrong Localizations</h3>
<p><em>Param$WrongLocalizations</em>: The percentage of wrong localizations in the MS search. Set to <code>0.01</code>. This parameter reflects the frequency of incorrect localization of PTM sites within identified peptides.</p>
</section>
</section>
<section id="filtering-ms-search-results" class="level2">
<h2 class="anchored" data-anchor-id="filtering-ms-search-results">Filtering MS Search Results</h2>
<section id="maximum-na-per-peptide" class="level3">
<h3 class="anchored" data-anchor-id="maximum-na-per-peptide">Maximum NA per Peptide</h3>
<p><em>Param$MaxNAPerPep</em>: The maximum number of NA values allowed per peptide. Set to <code>100</code>. This parameter filters out peptides with excessive missing values, ensuring more reliable quantitation.</p>
</section>
</section>
<section id="protein-summarization" class="level2">
<h2 class="anchored" data-anchor-id="protein-summarization">Protein Summarization</h2>
<section id="summarization-method" class="level3">
<h3 class="anchored" data-anchor-id="summarization-method">Summarization Method</h3>
<p><em>Param$ProtSummarization</em>: The method for protein summarization. Currently set to <code>medpolish</code>. This parameter determines how peptide-level data is aggregated to the protein level.</p>
</section>
<section id="minimum-number-of-unique-peptides" class="level3">
<h3 class="anchored" data-anchor-id="minimum-number-of-unique-peptides">Minimum Number of Unique Peptides</h3>
<p><em>Param$MinUniquePep</em>: The minimum number of unique peptides required for summarization. Set to <code>1</code>. This parameter ensures that proteins are summarized only if they have sufficient peptide evidence.</p>
</section>
</section>
<section id="statistical-testing-of-peptides-and-proteins" class="level2">
<h2 class="anchored" data-anchor-id="statistical-testing-of-peptides-and-proteins">Statistical Testing of Peptides and Proteins</h2>
<section id="paired-or-unpaired" class="level3">
<h3 class="anchored" data-anchor-id="paired-or-unpaired">Paired or Unpaired</h3>
<p><em>Param$StatPaired</em>: Indicates whether the statistical test is paired. Set to <code>FALSE</code>. This parameter determines the type of statistical test applied to the data, affecting the analysis of differential expression.</p>
</section>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>