<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content />

<meta name="date" content="2024-07-25" />

<title>PhosFake Vignette</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">PhosFake Vignette</h1>
<h4 class="author"></h4>
<h4 class="date">2024-07-25</h4>



<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>PhosFake is a simulation tool for generating synthetic proteomics
datasets. This vignette will guide you through the process of setting up
and running PhosFake simulations, as well as visualizing the
results.</p>
</div>
<div id="installation" class="section level2">
<h2>Installation</h2>
<p>PhosFake requires installation via the <code>devtools</code> package.
You can install PhosFake using the following command:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co"># Install PhosFake</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;veitveit/PhosFake&quot;</span>)</span></code></pre></div>
<pre><code>## Using GitHub PAT from the git credential store.</code></pre>
<pre><code>## Skipping install of &#39;PhosFake&#39; from a github remote, the SHA1 (5ccb1d6a) has not changed since last install.
##   Use `force = TRUE` to force installation</code></pre>
<p>And load the library</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># Load PhosFake</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="fu">library</span>(PhosFake)</span></code></pre></div>
</div>
<div id="configuration" class="section level2">
<h2>Configuration</h2>
<p>Set up the paths and directories for input and output data, and
configure the computational settings. This step ensures that the
necessary directories exist and the computational resources are
configured correctly. The fasta file path is set to the Proteomes folder
in the PhosFake package with the following proteomes:
<em>fasta_example.fasta, fasta_full_human.fasta, fasta_full_mouse.fasta,
fasta_full_yeast.fasta</em></p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># Configure the paths and settings</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>phosfake_config <span class="ot">&lt;-</span> <span class="fu">set_phosfake</span>(</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>  <span class="at">fastaFilePath =</span>  <span class="fu">system.file</span>(<span class="st">&quot;Proteomes&quot;</span>, <span class="at">package =</span> <span class="st">&quot;PhosFake&quot;</span>),</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>  <span class="at">resultFilePath =</span> <span class="st">&quot;SimulatedDataSets&quot;</span>,</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>  <span class="at">cores =</span> <span class="dv">2</span>,</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>  <span class="at">clusterType =</span> <span class="st">&quot;FORK&quot;</span>,</span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a>  <span class="at">runStatTests =</span> <span class="cn">TRUE</span>,</span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a>  <span class="at">calcAllBenchmarks =</span> <span class="cn">TRUE</span></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="set-simulation-parameters" class="section level2">
<h2>Set Simulation Parameters</h2>
<p>Generate a default list of testing parameters using the
<code>def_param</code> function. You can overwrite the default values
with those you want to test by changing the parameters directly, or by
loading a yaml file. An example for the yaml file with the default
parameters is available <a href="https://github.com/veitveit/PhosFake/blob/master/inst/config/parameters.yaml">here</a>.</p>
<p>Multiple values for each parameter can be combined to generate
different simulated datasets. This flexibility allows you to customize
the simulations according to your arrays of experimental needs.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># Generate default parameters</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>Param <span class="ot">&lt;-</span> <span class="fu">def_param</span>()</span></code></pre></div>
<pre><code>## 
## Attaching package: &#39;Hmisc&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     format.pval, units</code></pre>
<pre><code>## --------------------
## Ground truth generation parameters:
##  Param$paramGroundTruth = list 13 (2168 bytes)
## .  NumCond = integer 1= 2
## .  NumReps = integer 1= 3
## .  PathToFasta = character 1= fasta_full_yeast.fasta 
## .  PathToProteinList = logical 1= NA
## .  FracModProt = integer 1= 0
## .  PropModPerProt = integer 1= 0
## .  PTMTypes = logical 1= NA
## .  PTMTypesDist = logical 1= NA
## .  PTMTypesMass = logical 1= NA
## .  PTMMultipleLambda = logical 1= NA
## .  ModifiableResidues = logical 1= NA
## .  ModifiableResiduesDistr = logical 1= NA
## .  RemoveNonModFormFrac = integer 1= 0
## --------------------
## Proteoform abundance parameters:
##  Param$paramProteoformAb = list 9 (1656 bytes)
## .  QuantNoise = double 1= 0.5
## .  DiffRegFrac = double 1= 0.3
## .  DiffRegMax = integer 1= 1
## .  UserInputFoldChanges_NumRegProteoforms = logical 1= NA
## .  UserInputFoldChanges_RegulationFC = logical 1= NA
## .  ThreshNAProteoform = integer 1= 0
## .  AbsoluteQuanMean = double 1= 30.5
## .  AbsoluteQuanSD = double 1= 3.6
## .  ThreshNAQuantileProt = double 1= 0.01
## --------------------
## Digestion parameters:
##  Param$paramDigest = list 9 (1656 bytes)
## .  Enzyme = character 1= trypsin 
## .  PropMissedCleavages = double 1= 0.01
## .  MaxNumMissedCleavages = integer 1= 2
## .  PepMinLength = integer 1= 7
## .  PepMaxLength = integer 1= 30
## .  LeastAbundantLoss = integer 1= 0
## .  EnrichmentLoss = double 1= 0.2
## .  EnrichmentEfficiency = integer 1= 1
## .  EnrichmentNoise = double 1= 0.2
## --------------------
## MSRun parameters:
##  Param$paramMSRun = list 7 (1200 bytes)
## .  DetectabilityThreshold = double 1= 0.5
## .  PercDetectedVal = double 1= 0.5
## .  WeightDetectVal = double 1= 0.1
## .  MSNoise = double 1= 0.25
## .  WrongIDs = double 1= 0.01
## .  WrongLocalizations = integer 1= 0
## .  MaxNAPerPep = integer 1= 100
## --------------------
## Data analysis parameters:
##  Param$paramDataAnalysis = list 3 (712 bytes)
## .  ProtSummarization = character 1= medpolish 
## .  MinUniquePep = integer 1= 1
## .  StatPaired = logical 1= FALSE
## --------------------</code></pre>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="co"># Example of overwriting default values</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>Param<span class="sc">$</span>paramGroundTruth<span class="sc">$</span>PathToFasta <span class="ot">&lt;-</span> <span class="st">&quot;fasta_example.fasta&quot;</span></span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a><span class="co"># Param$paramGroundTruth$NumReps &lt;- c(3)</span></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a><span class="co"># Param$paramGroundTruth$NumCond &lt;- 5</span></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a><span class="co"># Param$paramGroundTruth$FracModProt &lt;- 0.5</span></span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a><span class="co"># Param$paramGroundTruth$PTMTypes &lt;- &quot;ph&quot;</span></span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a><span class="co"># Param$paramGroundTruth$PTMTypesMass &lt;- c(79.966331)</span></span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a><span class="co"># Param$paramGroundTruth$PTMTypesDist &lt;- c(1)</span></span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a><span class="co"># Param$paramGroundTruth$PTMMultipleLambda &lt;- c(0.1)</span></span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a><span class="co"># Param$paramGroundTruth$ModifiableResidues &lt;- list(c(&quot;S&quot;, &quot;T&quot;, &quot;Y&quot;))</span></span>
<span id="cb10-11"><a href="#cb10-11" tabindex="-1"></a><span class="co"># Param$paramGroundTruth$ModifiableResiduesDistr &lt;- list(c(0.86,0.13, 0.01))</span></span>
<span id="cb10-12"><a href="#cb10-12" tabindex="-1"></a>Param<span class="sc">$</span>paramGroundTruth<span class="sc">$</span>NumReps <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">3</span><span class="sc">:</span><span class="dv">5</span>)</span>
<span id="cb10-13"><a href="#cb10-13" tabindex="-1"></a><span class="co"># Param$paramProteoformAb$QuantNoise &lt;- seq(0.1, 0.9, 0.5)</span></span>
<span id="cb10-14"><a href="#cb10-14" tabindex="-1"></a><span class="co"># Param$paramProteoformAb$DiffRegFrac &lt;- c(0.1, 0.3, 0.5)</span></span>
<span id="cb10-15"><a href="#cb10-15" tabindex="-1"></a><span class="co"># Param$paramProteoformAb$DiffRegMax &lt;- seq(0.5, 2, 0.5)</span></span>
<span id="cb10-16"><a href="#cb10-16" tabindex="-1"></a><span class="co"># Param$paramDigest$Enzyme &lt;- &quot;trypsin&quot;</span></span>
<span id="cb10-17"><a href="#cb10-17" tabindex="-1"></a><span class="co"># Param$paramDigest$PropMissedCleavages &lt;- 0.01</span></span>
<span id="cb10-18"><a href="#cb10-18" tabindex="-1"></a><span class="co"># Param$paramDigest$MaxNumMissedCleavages &lt;- 4</span></span>
<span id="cb10-19"><a href="#cb10-19" tabindex="-1"></a><span class="co"># Param$paramDigest$PepMinLength &lt;- 7</span></span>
<span id="cb10-20"><a href="#cb10-20" tabindex="-1"></a><span class="co"># Param$paramDigest$PepMaxLength &lt;- 30</span></span>
<span id="cb10-21"><a href="#cb10-21" tabindex="-1"></a><span class="co"># Param$paramMSRun$PercDetectedPep &lt;- seq(0.1, 0.5, 0.1)</span></span>
<span id="cb10-22"><a href="#cb10-22" tabindex="-1"></a><span class="co"># Param$paramDataAnalysis$MinUniquePep &lt;- 100</span></span></code></pre></div>
</div>
<div id="running-simulations" class="section level2">
<h2>Running Simulations</h2>
<p>Run the simulations using the defined parameters and configuration.
This step generates the synthetic datasets based on the specified
parameters.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="co"># Run the simulations</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>allBs <span class="ot">&lt;-</span> <span class="fu">run_sims</span>(Param, phosfake_config)</span></code></pre></div>
<pre><code>## 
## Attaching package: &#39;dplyr&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:Hmisc&#39;:
## 
##     src, summarize</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     filter, lag</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     intersect, setdiff, setequal, union</code></pre>
<pre><code>## Total number of simulations to run:  3 
## #SAMPLE PREPARATION - Start
## 
##  + Importing data:
##   - File /home/veit/R/x86_64-pc-linux-gnu-library/4.4/PhosFake/Proteomes/fasta_example.fasta imported, containing 50 protein sequences.
##   - A total of 0 protein sequences have been removed due to unusual amino acids (B,J,O,U,X,Z) composition.
##   - A total of 0 duplicated protein accessions have been removed.
##   - Total number of remaining protein sequences: 50 
## 
##  + Creating modified and unmodified fractions:
##   - A total of 1 sequences are unmodifiable.
##   - The protein set cannot be fractionated, too low number of proteins or too low FracModProt %.
##   - Modified fraction: 0 proteins.
##   - Unmodified fraction: 50 proteins.
## 
## #SAMPLE PREPARATION - Finish
## 
## Add quan. distribution: Relative -&gt; absolute
## #PROTEOFORM DIGESTION - Start
## 
##  + Digestion input:
##   - A total number of 50 proteoforms, is proceed for proteolytic digestion.
##   - Unmodified fraction contains 50 proteoforms and modified fraction 0 proteoforms.
##   - Cleavage will be performed by trypsin with a maximum of 2 miss-cleavages, to create peptides of length 7 to 30 amino acids.
##   - All proteoforms are digested successfully!
## 
##  + Digestion output:
##   - A total number of 1448 peptides is generated.
##   - Unmodified fraction contains 1448 peptides and modified fraction 0 peptides.
##   - The amount of peptides with 0, 1, 2 miss-cleavages is 1432, 15, 1 respectively.
## 
## #PROTEOFORM DIGESTION - Finish
## 
## #PEPTIDE SUMMARIZATION - Start
## 
##  + Summarization input:
##   - A total number of 1448 peptides is proceed for summarization.
##   - Unique peptide IDs are generated.
##   - Isoleucine substitution to leucine is done.
##   - Peptide groups are generated.
##   - Peptide groups summarization is done.
##   - Remove 0 % of the least abundant peptides, which corresponds to 0 peptides.
## 
##  + Summarization output:
##   - A total number of  1409 summarized peptides is generated.
## 
## #PEPTIDE SUMMARIZATION - Finish
## 
## 
## #MS RUN SIMULATION - Start
## 
##  + Noise addition:
##   - The MS noise standard deviation is 0.25 .
##   - Noise added to all samples!
## 
##  + Detection limits:
##   - Removing peptides with detectability scores (PeptideRanger) lower than 0.5 .
##   - A total of 544 peptides is removed.
## 
##  + Removing peptide intensities:
##   - The percentage of remaining intensities is 50 % and the probabilities of selection depends on intensity values by 0.1 .
##   - A total of 4224 intensities is removed.
## 
##  + Addition of false identification:
##   - FDR selected is 1 % and corresponds to 14 peptides.
##   - FDR addition finished.
## 
##   - PTM re-location finished.
## 
## #MS RUN SIMULATION - Finish
## 
##  + Remove all modified peptides
##   - Remaining number of non-modified peptides: 1408 
##  + Protein summarisation
##   - Number of minimum unique peptide per protein: 1 
##   - Protein summarisation using the medpolish approach.
##   - Sorted protein table
##   - built protein index for faster summarization
##   - Initiated protein matrix for 51 protein groups
##   - Summarizing proteins, this can take a while
##   - Finished summarizing into 50 proteins</code></pre>
<pre><code>## Too few proteins or no statistical tests requested. 
##                                         Skipping this part.</code></pre>
<pre><code>## #SAMPLE PREPARATION - Start
## 
##  + Importing data:
##   - File /home/veit/R/x86_64-pc-linux-gnu-library/4.4/PhosFake/Proteomes/fasta_example.fasta imported, containing 50 protein sequences.
##   - A total of 0 protein sequences have been removed due to unusual amino acids (B,J,O,U,X,Z) composition.
##   - A total of 0 duplicated protein accessions have been removed.
##   - Total number of remaining protein sequences: 50 
## 
##  + Creating modified and unmodified fractions:
##   - A total of 1 sequences are unmodifiable.
##   - The protein set cannot be fractionated, too low number of proteins or too low FracModProt %.
##   - Modified fraction: 0 proteins.
##   - Unmodified fraction: 50 proteins.
## 
## #SAMPLE PREPARATION - Finish
## 
## Add quan. distribution: Relative -&gt; absolute
## #PROTEOFORM DIGESTION - Start
## 
##  + Digestion input:
##   - A total number of 50 proteoforms, is proceed for proteolytic digestion.
##   - Unmodified fraction contains 50 proteoforms and modified fraction 0 proteoforms.
##   - Cleavage will be performed by trypsin with a maximum of 2 miss-cleavages, to create peptides of length 7 to 30 amino acids.
##   - All proteoforms are digested successfully!
## 
##  + Digestion output:
##   - A total number of 1448 peptides is generated.
##   - Unmodified fraction contains 1448 peptides and modified fraction 0 peptides.
##   - The amount of peptides with 0, 1, 2 miss-cleavages is 1432, 15, 1 respectively.
## 
## #PROTEOFORM DIGESTION - Finish
## 
## #PEPTIDE SUMMARIZATION - Start
## 
##  + Summarization input:
##   - A total number of 1448 peptides is proceed for summarization.
##   - Unique peptide IDs are generated.
##   - Isoleucine substitution to leucine is done.
##   - Peptide groups are generated.
##   - Peptide groups summarization is done.
##   - Remove 0 % of the least abundant peptides, which corresponds to 0 peptides.
## 
##  + Summarization output:
##   - A total number of  1409 summarized peptides is generated.
## 
## #PEPTIDE SUMMARIZATION - Finish
## 
## 
## #MS RUN SIMULATION - Start
## 
##  + Noise addition:
##   - The MS noise standard deviation is 0.25 .
##   - Noise added to all samples!
## 
##  + Detection limits:
##   - Removing peptides with detectability scores (PeptideRanger) lower than 0.5 .
##   - A total of 544 peptides is removed.
## 
##  + Removing peptide intensities:
##   - The percentage of remaining intensities is 50 % and the probabilities of selection depends on intensity values by 0.1 .
##   - A total of 5632 intensities is removed.
## 
##  + Addition of false identification:
##   - FDR selected is 1 % and corresponds to 14 peptides.
##   - FDR addition finished.
## 
##   - PTM re-location finished.
## 
## #MS RUN SIMULATION - Finish
## 
##  + Remove all modified peptides
##   - Remaining number of non-modified peptides: 1408 
##  + Protein summarisation
##   - Number of minimum unique peptide per protein: 1 
##   - Protein summarisation using the medpolish approach.
##   - Sorted protein table
##   - built protein index for faster summarization
##   - Initiated protein matrix for 51 protein groups
##   - Summarizing proteins, this can take a while
##   - Finished summarizing into 50 proteins</code></pre>
<pre><code>## Too few proteins or no statistical tests requested. 
##                                         Skipping this part.</code></pre>
<pre><code>## #SAMPLE PREPARATION - Start
## 
##  + Importing data:
##   - File /home/veit/R/x86_64-pc-linux-gnu-library/4.4/PhosFake/Proteomes/fasta_example.fasta imported, containing 50 protein sequences.
##   - A total of 0 protein sequences have been removed due to unusual amino acids (B,J,O,U,X,Z) composition.
##   - A total of 0 duplicated protein accessions have been removed.
##   - Total number of remaining protein sequences: 50 
## 
##  + Creating modified and unmodified fractions:
##   - A total of 1 sequences are unmodifiable.
##   - The protein set cannot be fractionated, too low number of proteins or too low FracModProt %.
##   - Modified fraction: 0 proteins.
##   - Unmodified fraction: 50 proteins.
## 
## #SAMPLE PREPARATION - Finish
## 
## Add quan. distribution: Relative -&gt; absolute
## #PROTEOFORM DIGESTION - Start
## 
##  + Digestion input:
##   - A total number of 50 proteoforms, is proceed for proteolytic digestion.
##   - Unmodified fraction contains 50 proteoforms and modified fraction 0 proteoforms.
##   - Cleavage will be performed by trypsin with a maximum of 2 miss-cleavages, to create peptides of length 7 to 30 amino acids.
##   - All proteoforms are digested successfully!
## 
##  + Digestion output:
##   - A total number of 1448 peptides is generated.
##   - Unmodified fraction contains 1448 peptides and modified fraction 0 peptides.
##   - The amount of peptides with 0, 1, 2 miss-cleavages is 1432, 15, 1 respectively.
## 
## #PROTEOFORM DIGESTION - Finish
## 
## #PEPTIDE SUMMARIZATION - Start
## 
##  + Summarization input:
##   - A total number of 1448 peptides is proceed for summarization.
##   - Unique peptide IDs are generated.
##   - Isoleucine substitution to leucine is done.
##   - Peptide groups are generated.
##   - Peptide groups summarization is done.
##   - Remove 0 % of the least abundant peptides, which corresponds to 0 peptides.
## 
##  + Summarization output:
##   - A total number of  1409 summarized peptides is generated.
## 
## #PEPTIDE SUMMARIZATION - Finish
## 
## 
## #MS RUN SIMULATION - Start
## 
##  + Noise addition:
##   - The MS noise standard deviation is 0.25 .
##   - Noise added to all samples!
## 
##  + Detection limits:
##   - Removing peptides with detectability scores (PeptideRanger) lower than 0.5 .
##   - A total of 544 peptides is removed.
## 
##  + Removing peptide intensities:
##   - The percentage of remaining intensities is 50 % and the probabilities of selection depends on intensity values by 0.1 .
##   - A total of 7040 intensities is removed.
## 
##  + Addition of false identification:
##   - FDR selected is 1 % and corresponds to 14 peptides.
##   - FDR addition finished.
## 
##   - PTM re-location finished.
## 
## #MS RUN SIMULATION - Finish
## 
##  + Remove all modified peptides
##   - Remaining number of non-modified peptides: 1408 
##  + Protein summarisation
##   - Number of minimum unique peptide per protein: 1 
##   - Protein summarisation using the medpolish approach.
##   - Sorted protein table
##   - built protein index for faster summarization
##   - Initiated protein matrix for 51 protein groups
##   - Summarizing proteins, this can take a while
##   - Finished summarizing into 49 proteins</code></pre>
<pre><code>## Too few proteins or no statistical tests requested. 
##                                         Skipping this part.</code></pre>
</div>
<div id="retrieving-results" class="section level2">
<h2>Retrieving Results</h2>
<p>Get the results of an individual simulation. This allows you to
inspect the parameters and outputs of a specific run. Retrieving results
is useful for detailed analysis of a particular simulation.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="co"># Retrieve results}</span></span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">get_simulation</span>(allBs[[<span class="dv">1</span>]]<span class="sc">$</span>Param, phosfake_config)</span></code></pre></div>
<pre><code>## No simulation found with these parameters.</code></pre>
</div>
<div id="retrieving-benchmarks" class="section level2">
<h2>Retrieving Benchmarks</h2>
<p>Create a matrix of benchmarks from all simulations and save it as a
CSV file. This file contains various metrics and results from the
simulations. Generating benchmarks helps in comparing different
simulations and assessing their performance.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a><span class="co"># Generate the benchmark matrix</span></span>
<span id="cb24-2"><a href="#cb24-2" tabindex="-1"></a>benchmarks <span class="ot">&lt;-</span> <span class="fu">matrix_benchmarks</span>(allBs, phosfake_config)</span>
<span id="cb24-3"><a href="#cb24-3" tabindex="-1"></a><span class="fu">write.csv</span>(benchmarks, <span class="at">file =</span> <span class="fu">paste0</span>(phosfake_config<span class="sc">$</span>resultFilePath, <span class="st">&quot;/allBenchmarks.csv&quot;</span>))</span></code></pre></div>
</div>
<div id="visualizing-results" class="section level2">
<h2>Visualizing Results</h2>
<p>Visualize the results of the benchmarks to analyze the performance
and outcomes of the simulations. Visualization aids in interpreting the
data and identifyinsg trends or anomalies</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a><span class="co"># visualize the benchmarks and parameters of the second simulation</span></span>
<span id="cb25-2"><a href="#cb25-2" tabindex="-1"></a><span class="fu">visualize_benchmarks</span>(benchmarks, <span class="dv">2</span>)</span></code></pre></div>
<pre><code>## No benchmarks available for this simulation.</code></pre>
<pre><code>## NULL</code></pre>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
