<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="">
<meta name="dcterms.date" content="2024-07-25">

<title>PhosFake Vignette</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="Vignette_files/libs/clipboard/clipboard.min.js"></script>
<script src="Vignette_files/libs/quarto-html/quarto.js"></script>
<script src="Vignette_files/libs/quarto-html/popper.min.js"></script>
<script src="Vignette_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="Vignette_files/libs/quarto-html/anchor.min.js"></script>
<link href="Vignette_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="Vignette_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="Vignette_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="Vignette_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="Vignette_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">PhosFake Vignette</h1>
</div>



<div class="quarto-title-meta">

    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">July 25, 2024</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">Introduction</h2>
<p>PhosFake is a simulation tool for generating synthetic proteomics datasets. This vignette will guide you through the process of setting up and running PhosFake simulations, as well as visualizing the results.</p>
</section>
<section id="installation" class="level2">
<h2 class="anchored" data-anchor-id="installation">Installation</h2>
<p>PhosFake requires installation via the <code>devtools</code> package. You can install PhosFake using the following command:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Install PhosFake</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"veitveit/PhosFake"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Using GitHub PAT from the git credential store.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Downloading GitHub repo veitveit/PhosFake@HEAD</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Skipping 4 packages not available: SummarizedExperiment, limma, qvalue, preprocessCore</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>── R CMD build ─────────────────────────────────────────────────────────────────
* checking for file ‘/tmp/Rtmplkja3L/remotesc997b38133bef/veitveit-PhosFake-b797848/DESCRIPTION’ ... OK
* preparing ‘PhosFake’:
* checking DESCRIPTION meta-information ... OK
* excluding invalid files
Subdirectory 'man' contains invalid file names:
  ‘Benchmarks.html’ ‘Benchmarks.qmd’ ‘Parameters.html’ ‘Parameters.qmd’
* checking for LF line-endings in source and make files and shell scripts
* checking for empty or unneeded directories
Omitted ‘LazyData’ from DESCRIPTION
* building ‘PhosFake_0.1.0.tar.gz’</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Installing package into '/home/veit/R/x86_64-pc-linux-gnu-library/4.4'
(as 'lib' is unspecified)</code></pre>
</div>
</div>
<p>And load the library</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load PhosFake</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(PhosFake)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="configuration" class="level2">
<h2 class="anchored" data-anchor-id="configuration">Configuration</h2>
<p>Set up the paths and directories for input and output data, and configure the computational settings. This step ensures that the necessary directories exist and the computational resources are configured correctly. The fasta file path is set to the Proteomes folder in the PhosFake package with the following proteomes: <em>fasta_example.fasta, fasta_full_human.fasta, fasta_full_mouse.fasta, fasta_full_yeast.fasta</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Configure the paths and settings</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>phosfake_config <span class="ot">&lt;-</span> <span class="fu">set_phosfake</span>(</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">fastaFilePath =</span>  <span class="fu">system.file</span>(<span class="st">"Proteomes"</span>, <span class="at">package =</span> <span class="st">"PhosFake"</span>),</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">resultFilePath =</span> <span class="st">"SimulatedDataSets"</span>,</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">cores =</span> <span class="dv">2</span>,</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">clusterType =</span> <span class="st">"FORK"</span>,</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">calcAllBenchmarks =</span> <span class="cn">TRUE</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="set-simulation-parameters" class="level2">
<h2 class="anchored" data-anchor-id="set-simulation-parameters">Set Simulation Parameters</h2>
<p>Generate a default list of testing parameters using the <code>def_param</code> function. You can overwrite the default values with those you want to test by changing the parameters directly, or by loading a yaml file. An example for the yaml file with the default parameters is available <a href="https://github.com/veitveit/PhosFake/blob/master/inst/config/parameters.yaml">here</a>.</p>
<p>Multiple values for each parameter can be combined to generate different simulated datasets. This flexibility allows you to customize the simulations according to your arrays of experimental needs.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate default parameters</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>Param <span class="ot">&lt;-</span> <span class="fu">def_param</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'Hmisc'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    format.pval, units</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>--------------------
Ground truth generation parameters:
 Param$paramGroundTruth = list 13 (2168 bytes)
.  NumCond = integer 1= 2
.  NumReps = integer 1= 3
.  PathToFasta = character 1= fasta_full_yeast.fasta 
.  PathToProteinList = logical 1= NA
.  FracModProt = integer 1= 0
.  FracModPerProt = integer 1= 0
.  PTMTypes = logical 1= NA
.  PTMTypesDist = logical 1= NA
.  PTMTypesMass = logical 1= NA
.  PTMMultipleLambda = logical 1= NA
.  ModifiableResidues = logical 1= NA
.  ModifiableResiduesDistr = logical 1= NA
.  ...   and 1 more
--------------------
Proteoform abundance parameters:
 Param$paramProteoformAb = list 9 (1656 bytes)
.  QuantNoise = double 1= 0.5
.  DiffRegFrac = double 1= 0.3
.  DiffRegMax = integer 1= 1
.  UserInputFoldChanges_NumRegProteoforms = logical 1= NA
.  UserInputFoldChanges_RegulationFC = logical 1= NA
.  ThreshNAProteoform = integer 1= 0
.  AbsoluteQuanMean = double 1= 30.5
.  AbsoluteQuanSD = double 1= 3.6
.  ThreshNAQuantileProt = double 1= 0.01
--------------------
Digestion parameters:
 Param$paramDigest = list 9 (1656 bytes)
.  Enzyme = character 1= trypsin 
.  PropMissedCleavages = double 1= 0.01
.  MaxNumMissedCleavages = integer 1= 2
.  PepMinLength = integer 1= 7
.  PepMaxLength = integer 1= 30
.  LeastAbundantLoss = integer 1= 0
.  EnrichmentLoss = double 1= 0.2
.  EnrichmentEfficiency = integer 1= 1
.  EnrichmentNoise = double 1= 0.2
--------------------
MSRun parameters:
 Param$paramMSRun = list 7 (1200 bytes)
.  DetectabilityThreshold = double 1= 0.5
.  PercDetectedVal = double 1= 0.5
.  WeightDetectVal = double 1= 0.1
.  MSNoise = double 1= 0.25
.  WrongIDs = double 1= 0.01
.  WrongLocalizations = integer 1= 0
.  MaxNAPerPep = integer 1= 100
--------------------
Data analysis parameters:
 Param$paramDataAnalysis = list 3 (712 bytes)
.  ProtSummarization = character 1= medpolish 
.  MinUniquePep = integer 1= 1
.  StatPaired = logical 1= FALSE
--------------------</code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Example of overwriting default values</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>Param<span class="sc">$</span>paramGroundTruth<span class="sc">$</span>PathToFasta <span class="ot">&lt;-</span> <span class="st">"fasta_example.fasta"</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Param$paramGroundTruth$NumReps &lt;- c(3)</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Param$paramGroundTruth$NumCond &lt;- 5</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Param$paramGroundTruth$FracModProt &lt;- 0.5</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Param$paramGroundTruth$PTMTypes &lt;- "ph"</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Param$paramGroundTruth$PTMTypesMass &lt;- c(79.966331)</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Param$paramGroundTruth$PTMTypesDist &lt;- c(1)</span></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Param$paramGroundTruth$PTMMultipleLambda &lt;- c(0.1)</span></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Param$paramGroundTruth$ModifiableResidues &lt;- list(c("S", "T", "Y"))</span></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Param$paramGroundTruth$ModifiableResiduesDistr &lt;- list(c(0.86,0.13, 0.01))</span></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>Param<span class="sc">$</span>paramGroundTruth<span class="sc">$</span>NumReps <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">3</span><span class="sc">:</span><span class="dv">5</span>)</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Param$paramProteoformAb$QuantNoise &lt;- seq(0.1, 0.9, 0.5)</span></span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Param$paramProteoformAb$DiffRegFrac &lt;- c(0.1, 0.3, 0.5)</span></span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Param$paramProteoformAb$DiffRegMax &lt;- seq(0.5, 2, 0.5)</span></span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Param$paramDigest$Enzyme &lt;- "trypsin"</span></span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Param$paramDigest$PropMissedCleavages &lt;- 0.01</span></span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Param$paramDigest$MaxNumMissedCleavages &lt;- 4</span></span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Param$paramDigest$PepMinLength &lt;- 7</span></span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Param$paramDigest$PepMaxLength &lt;- 30</span></span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Param$paramMSRun$PercDetectedPep &lt;- seq(0.1, 0.5, 0.1)</span></span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Param$paramDataAnalysis$MinUniquePep &lt;- 100</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="running-simulations" class="level2">
<h2 class="anchored" data-anchor-id="running-simulations">Running Simulations</h2>
<p>Run the simulations using the defined parameters and configuration. This step generates the synthetic datasets based on the specified parameters.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Run the simulations</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>allBs <span class="ot">&lt;-</span> <span class="fu">run_sims</span>(Param, phosfake_config)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'dplyr'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:Hmisc':

    src, summarize</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    filter, lag</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Total number of simulations to run:  3 
#SAMPLE PREPARATION - Start

 + Importing data:
  - File /home/veit/R/x86_64-pc-linux-gnu-library/4.4/PhosFake/Proteomes/fasta_example.fasta imported, containing 50 protein sequences.
  - A total of 0 protein sequences have been removed due to unusual amino acids (B,J,O,U,X,Z) composition.
  - A total of 0 duplicated protein accessions have been removed.
  - Total number of remaining protein sequences: 50 

 + Creating modified and unmodified fractions:
  - A total of 1 sequences are unmodifiable.
  - The protein set cannot be fractionated, too low number of proteins or too low FracModProt %.
  - Modified fraction: 0 proteins.
  - Unmodified fraction: 50 proteins.

#SAMPLE PREPARATION - Finish

Add quan. distribution: Relative -&gt; absolute
#PROTEOFORM DIGESTION - Start

 + Digestion input:
  - A total number of 50 proteoforms, is proceed for proteolytic digestion.
  - Unmodified fraction contains 50 proteoforms and modified fraction 0 proteoforms.
  - Cleavage will be performed by trypsin with a maximum of 2 miss-cleavages, to create peptides of length 7 to 30 amino acids.
  - All proteoforms are digested successfully!

 + Digestion output:
  - A total number of 1448 peptides is generated.
  - Unmodified fraction contains 1448 peptides and modified fraction 0 peptides.
  - The amount of peptides with 0, 1, 2 miss-cleavages is 1432, 15, 1 respectively.

#PROTEOFORM DIGESTION - Finish

#PEPTIDE SUMMARIZATION - Start

 + Summarization input:
  - A total number of 1448 peptides is proceed for summarization.
  - Unique peptide IDs are generated.
  - Isoleucine substitution to leucine is done.
  - Peptide groups are generated.
  - Peptide groups summarization is done.
  - Remove 0 % of the least abundant peptides, which corresponds to 0 peptides.

 + Summarization output:
  - A total number of  1409 summarized peptides is generated.

#PEPTIDE SUMMARIZATION - Finish


#MS RUN SIMULATION - Start

 + Noise addition:
  - The MS noise standard deviation is 0.25 .
  - Noise added to all samples!

 + Detection limits:
  - Removing peptides with detectability scores (PeptideRanger) lower than 0.5 .
  - A total of 542 peptides is removed.

 + Removing peptide intensities:
  - The percentage of remaining intensities is 50 % and the probabilities of selection depends on intensity values by 0.1 .
  - A total of 4224 intensities is removed.

 + Addition of false identification:
  - FDR selected is 1 % and corresponds to 14 peptides.
  - FDR addition finished.

  - PTM re-location finished.

#MS RUN SIMULATION - Finish

 + Remove all modified peptides
  - Remaining number of non-modified peptides: 1408 
 + Protein summarisation
  - Number of minimum unique peptide per protein: 1 
  - Protein summarisation using the medpolish approach.
  - Sorted protein table
  - built protein index for faster summarization
  - Initiated protein matrix for 51 protein groups
  - Summarizing proteins, this can take a while
  - Finished summarizing into 49 proteins</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Too few proteins!!!</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>#SAMPLE PREPARATION - Start

 + Importing data:
  - File /home/veit/R/x86_64-pc-linux-gnu-library/4.4/PhosFake/Proteomes/fasta_example.fasta imported, containing 50 protein sequences.
  - A total of 0 protein sequences have been removed due to unusual amino acids (B,J,O,U,X,Z) composition.
  - A total of 0 duplicated protein accessions have been removed.
  - Total number of remaining protein sequences: 50 

 + Creating modified and unmodified fractions:
  - A total of 1 sequences are unmodifiable.
  - The protein set cannot be fractionated, too low number of proteins or too low FracModProt %.
  - Modified fraction: 0 proteins.
  - Unmodified fraction: 50 proteins.

#SAMPLE PREPARATION - Finish

Add quan. distribution: Relative -&gt; absolute
#PROTEOFORM DIGESTION - Start

 + Digestion input:
  - A total number of 50 proteoforms, is proceed for proteolytic digestion.
  - Unmodified fraction contains 50 proteoforms and modified fraction 0 proteoforms.
  - Cleavage will be performed by trypsin with a maximum of 2 miss-cleavages, to create peptides of length 7 to 30 amino acids.
  - All proteoforms are digested successfully!

 + Digestion output:
  - A total number of 1448 peptides is generated.
  - Unmodified fraction contains 1448 peptides and modified fraction 0 peptides.
  - The amount of peptides with 0, 1, 2 miss-cleavages is 1432, 15, 1 respectively.

#PROTEOFORM DIGESTION - Finish

#PEPTIDE SUMMARIZATION - Start

 + Summarization input:
  - A total number of 1448 peptides is proceed for summarization.
  - Unique peptide IDs are generated.
  - Isoleucine substitution to leucine is done.
  - Peptide groups are generated.
  - Peptide groups summarization is done.
  - Remove 0 % of the least abundant peptides, which corresponds to 0 peptides.

 + Summarization output:
  - A total number of  1409 summarized peptides is generated.

#PEPTIDE SUMMARIZATION - Finish


#MS RUN SIMULATION - Start

 + Noise addition:
  - The MS noise standard deviation is 0.25 .
  - Noise added to all samples!

 + Detection limits:
  - Removing peptides with detectability scores (PeptideRanger) lower than 0.5 .
  - A total of 544 peptides is removed.

 + Removing peptide intensities:
  - The percentage of remaining intensities is 50 % and the probabilities of selection depends on intensity values by 0.1 .
  - A total of 5632 intensities is removed.

 + Addition of false identification:
  - FDR selected is 1 % and corresponds to 14 peptides.
  - FDR addition finished.

  - PTM re-location finished.

#MS RUN SIMULATION - Finish

 + Remove all modified peptides
  - Remaining number of non-modified peptides: 1408 
 + Protein summarisation
  - Number of minimum unique peptide per protein: 1 
  - Protein summarisation using the medpolish approach.
  - Sorted protein table
  - built protein index for faster summarization
  - Initiated protein matrix for 51 protein groups
  - Summarizing proteins, this can take a while
  - Finished summarizing into 49 proteins</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Too few proteins!!!</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>#SAMPLE PREPARATION - Start

 + Importing data:
  - File /home/veit/R/x86_64-pc-linux-gnu-library/4.4/PhosFake/Proteomes/fasta_example.fasta imported, containing 50 protein sequences.
  - A total of 0 protein sequences have been removed due to unusual amino acids (B,J,O,U,X,Z) composition.
  - A total of 0 duplicated protein accessions have been removed.
  - Total number of remaining protein sequences: 50 

 + Creating modified and unmodified fractions:
  - A total of 1 sequences are unmodifiable.
  - The protein set cannot be fractionated, too low number of proteins or too low FracModProt %.
  - Modified fraction: 0 proteins.
  - Unmodified fraction: 50 proteins.

#SAMPLE PREPARATION - Finish

Add quan. distribution: Relative -&gt; absolute
#PROTEOFORM DIGESTION - Start

 + Digestion input:
  - A total number of 50 proteoforms, is proceed for proteolytic digestion.
  - Unmodified fraction contains 50 proteoforms and modified fraction 0 proteoforms.
  - Cleavage will be performed by trypsin with a maximum of 2 miss-cleavages, to create peptides of length 7 to 30 amino acids.
  - All proteoforms are digested successfully!

 + Digestion output:
  - A total number of 1448 peptides is generated.
  - Unmodified fraction contains 1448 peptides and modified fraction 0 peptides.
  - The amount of peptides with 0, 1, 2 miss-cleavages is 1432, 15, 1 respectively.

#PROTEOFORM DIGESTION - Finish

#PEPTIDE SUMMARIZATION - Start

 + Summarization input:
  - A total number of 1448 peptides is proceed for summarization.
  - Unique peptide IDs are generated.
  - Isoleucine substitution to leucine is done.
  - Peptide groups are generated.
  - Peptide groups summarization is done.
  - Remove 0 % of the least abundant peptides, which corresponds to 0 peptides.

 + Summarization output:
  - A total number of  1409 summarized peptides is generated.

#PEPTIDE SUMMARIZATION - Finish


#MS RUN SIMULATION - Start

 + Noise addition:
  - The MS noise standard deviation is 0.25 .
  - Noise added to all samples!

 + Detection limits:
  - Removing peptides with detectability scores (PeptideRanger) lower than 0.5 .
  - A total of 544 peptides is removed.

 + Removing peptide intensities:
  - The percentage of remaining intensities is 50 % and the probabilities of selection depends on intensity values by 0.1 .
  - A total of 7040 intensities is removed.

 + Addition of false identification:
  - FDR selected is 1 % and corresponds to 14 peptides.
  - FDR addition finished.

  - PTM re-location finished.

#MS RUN SIMULATION - Finish

 + Remove all modified peptides
  - Remaining number of non-modified peptides: 1408 
 + Protein summarisation
  - Number of minimum unique peptide per protein: 1 
  - Protein summarisation using the medpolish approach.
  - Sorted protein table
  - built protein index for faster summarization
  - Initiated protein matrix for 51 protein groups
  - Summarizing proteins, this can take a while
  - Finished summarizing into 49 proteins</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Too few proteins!!!</code></pre>
</div>
</div>
</section>
<section id="retrieving-results" class="level2">
<h2 class="anchored" data-anchor-id="retrieving-results">Retrieving Results</h2>
<p>Get the results of an individual simulation. This allows you to inspect the parameters and outputs of a specific run. Retrieving results is useful for detailed analysis of a particular simulation.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Retrieve results}</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">get_simulation</span>(allBs[[<span class="dv">1</span>]]<span class="sc">$</span>Param, phosfake_config)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="retrieving-benchmarks" class="level2">
<h2 class="anchored" data-anchor-id="retrieving-benchmarks">Retrieving Benchmarks</h2>
<p>Create a matrix of benchmarks from all simulations and save it as a CSV file. This file contains various metrics and results from the simulations. Generating benchmarks helps in comparing different simulations and assessing their performance.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate the benchmark matrix</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>benchmarks <span class="ot">&lt;-</span> <span class="fu">matrix_benchmarks</span>(allBs, phosfake_config)</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(benchmarks, <span class="at">file =</span> <span class="fu">paste0</span>(phosfake_config<span class="sc">$</span>resultFilePath, <span class="st">"/allBenchmarks.csv"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="visualizing-results" class="level2">
<h2 class="anchored" data-anchor-id="visualizing-results">Visualizing Results</h2>
<p>Visualize the results of the benchmarks to analyze the performance and outcomes of the simulations. Visualization aids in interpreting the data and identifyinsg trends or anomalies</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># visualize the benchmarks and parameters of the second simulation</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">visualize_benchmarks</span>(benchmarks, <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>No benchmarks available for this simulation.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>NULL</code></pre>
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>